gflownet/utils/batch.py

```diff
@@ -0,0 +1,179 @@
+from collections import defaultdict
+from copy import deepcopy
+
+import numpy as np
+import torch
+
+from gflownet.utils.common import tbool, tfloat, tint, tlong
+
+
+class Batch:
+    """
+    Important note: one env should correspond to only one trajectory, all env_id should be unique
+    """
+
+    def __init__(self, loss, device, float):
+        self.loss = loss
+        self.device = device
+        self.float = float
+        self.envs = dict()
+        self.state = []
+        self.action = []
+        self.done = []
+        self.env_id = []
+        self.mask_invalid_actions_forward = []
+        self.mask_invalid_actions_backward = []
+        self.parents = []
+        self.parents_actions = []
+        self.step = []
+
+    def __len__(self):
+        return len(self.state)
+
+    def add_to_batch(self, envs, actions, valids, train=True):
+        for env, action, valid in zip(envs, actions, valids):
+            self.envs.update({env.id: env})
```

Comment:
Not that this is wrong, but is there any benefit of using update with a single-element dict as an argument?

Reply:
Sorry, I don't get what you mean, could you elaborate?

Reply:
I think he means is there a benefit to doing this : 
`self.envs.update({env.id: env})`
rather than this : 
`self.envs[env.id] = env`

To my knowledge, most people consider the second one to be more readable. Aside from that, it is also faster (see : https://stackoverflow.com/questions/15456158/python-dict-update-vs-subscript-to-add-a-single-key-value-pair) but I don't think that speed is a big issue here.

Reply:
Although, some people will argue that using the update function is more OO

Reply:
I was not aware of this "debate". I also tend to use `update` because you need it to update more than one key in one go, so it seems more consistent to me to do the same with just one key. Anyway, speed matters!

So, should we change it to `self.envs[env.id] = env`?